### PR TITLE
Support continuous profiling feature

### PR DESCRIPTION
**Pull request by mrproliu**

### CODE REVIEW

File: ebpf/profiling/Continuous.proto [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+message ContinuousProfilingReport {
+    // over threshold process entity
+    string layer = 1;
+    string serviceName = 2;
+    string instanceName = 3;
+    string processName = 4;
+
+    // reached thresholds
+    repeated ContinuousProfilingThreshold reaches = 5;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:59]:
```suggestion
    repeated ContinuousProfilingThreshold causes = 5;
```

File: ebpf/profiling/Continuous.proto [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    int32 duration = 6;
+    // target profiling task
+    oneof targetTask {
+        ContinuousOnCPUProfilingTask onCPU = 7;
+        ContinuousOffCPUProfilingTask offCPU = 8;
+        ContinuousNetworkProfilingTask networkCPU = 9;
+    }
+}
+
+message ContinuousProfilingThreshold {
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:71]:
```suggestion
message ContinuousProfilingCause {
```

File: ebpf/profiling/Continuous.proto [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    string type = 1;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:72]: What does this type mean?

File: ebpf/profiling/Continuous.proto [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    oneof threshold {
+        ContinuousProfilingSingleValueThreshold singleValue = 2;
+        ContinuousProfilingURIThreshold uri = 3;
+    }
+}
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:77]: All `threshold`s should be renamed as `Cause` when profiling results report.

File: ebpf/profiling/Continuous.proto [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    repeated ContinuousProfilingThreshold reaches = 5;
+
+    // target profiling task execute duration
+    int32 duration = 6;
+    // target profiling task
+    oneof targetTask {
+        ContinuousOnCPUProfilingTask onCPU = 7;
+        ContinuousOffCPUProfilingTask offCPU = 8;
+        ContinuousNetworkProfilingTask networkCPU = 9;
+    }
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:68]: Why do we need these three and network profiling even with some settings? Who pushed these network profiling settings?

File: ebpf/profiling/Continuous.proto [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    // reached thresholds
+    repeated ContinuousProfilingThreshold reaches = 5;
+
+    // target profiling task execute duration
+    int32 duration = 6;
+    // target profiling task
+    oneof targetTask {
+        ContinuousOnCPUProfilingTask onCPU = 7;
+        ContinuousOffCPUProfilingTask offCPU = 8;
+        ContinuousNetworkProfilingTask networkCPU = 9;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:67]:
```suggestion
        ContinuousNetworkProfilingTask network = 9;
```

File: ebpf/profiling/Continuous.proto (hunk 3 above)
Comment (mrproliu) [reply ebpf/profiling/Continuous.proto:72]: It typically aligns with the ”Monitoring type“ defined in the Command, and a String is employed to allow for future expansion without the need for modifying the protocol.

Comment (wu-sheng) [reply ebpf/profiling/Continuous.proto:72]: Please be clear in the comment.

File: ebpf/profiling/Continuous.proto (hunk 5 above)
Comment (mrproliu) [reply ebpf/profiling/Continuous.proto:68]: These three types of data are all reported by the rover. for the network profiling, the settings are set by the rover. It should perform sampling analysis on specifically set URIs from the policy.

Comment (wu-sheng) [reply ebpf/profiling/Continuous.proto:68]: If those settings are static configurations on the rover side, why do we need to report them? We have collected data no matter what the sampling rules are.

Comment (wu-sheng) [reply ebpf/profiling/Continuous.proto:68]: If you want to report metrics to show values that matched the policy(s), it is better you use meter channel to do that in an official way. The auto-profiling should be an event when triggered.

Comment (mrproliu) [reply ebpf/profiling/Continuous.proto:68]: > If those settings are static configurations on the rover side, why do we need to report them? We have collected data no matter what the sampling rules are.

It is not static. The sampling settings should be decided by which URI(or Regex) matches the threshold.

Comment (wu-sheng) [reply ebpf/profiling/Continuous.proto:68]: I mean these
```
message ContinuousNetworkProfilingTaskSampling {
    string uriRegex = 1;
    int32 minDuration = 2;
    bool when4xx = 3;
    bool when5xx = 4;
    bool completeRequest = 5;
    bool completeResponse = 6;
}
```

I can't see these are as auto profiling parameters. They should be static, right?

Comment (mrproliu) [reply ebpf/profiling/Continuous.proto:68]: > If you want to report metrics to show values that matched the policy(s), it is better you use meter channel to do that in an official way. The auto-profiling should be an event when triggered.

Yes, the metrics would be using the native meter protocol. The cases only work for taking a snapshot for building an event on the OAP side.

Comment (mrproliu) [reply ebpf/profiling/Continuous.proto:68]: > I can't see these are as auto profiling parameters. They should be static, right?

Ok, the `uriRegex` should be reported, and other fields should be removed in this protocol.

Comment (wu-sheng) [reply ebpf/profiling/Continuous.proto]: I think you should decouple the two things. 
1. Auto profiling triggered, which should include causes. Only URI cause may include URIRegex. But basically, this should only refer to an existing policy. Do we have a policy ID?
2. Other profiling data should be reported with a task(UUID or something) right? 

Reporting causes repeatedly seems duplicate and not helpful

File: ebpf/profiling/Continuous.proto (hunk 3 above)
Comment (mrproliu) [reply ebpf/profiling/Continuous.proto:74]: updated.

File: ebpf/profiling/Continuous.proto (hunk 5 above)
Comment (mrproliu) [reply ebpf/profiling/Continuous.proto]: > Auto profiling triggered, which should include causes. Only URI cause may include URIRegex. But basically, this should only refer to an existing policy. Do we have a policy ID?

For now, we haven't the policy ID. If we add all policies with an ID may be too complex for this feature. Maybe I could copy the policy content into the report protocol? 

> Other profiling data should be reported with a task(UUID or something) right?

If the policy is reached, then the rover will call the report protocol to get a TaskID(which generates on the OAP). The report should be ignored when profiling or matching the silence policy.

Comment (wu-sheng) [reply ebpf/profiling/Continuous.proto]: > For now, we haven't the policy ID. If we add all policies with an ID may be too complex for this feature. Maybe I could copy the policy content into the report protocol?

If you don't. I can't see why we plan to link the profiling to the policy. The policy could change with time. The link will lost the point.

Comment (wu-sheng) [reply ebpf/profiling/Continuous.proto]: > For now, we haven't the policy ID. If we add all policies with an ID may be too complex for this feature. Maybe I could copy the policy content into the report protocol?

If you don't. I can't see why we plan to link the profiling to the policy. The policy could change with time. The link will lose the point.

File: ebpf/profiling/Continuous.proto [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+message ContinuousProfilingSingleValueCause {
+    double threshold = 1;
+    double current = 2;
+}
+
+message ContinuousProfilingURICause {
+    string uriRegex = 1;
+    string uriPath = 2;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:88]: OneOf?

File: ebpf/profiling/Continuous.proto [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// eBPF Network profiling task
+message ContinuousNetworkProfilingTask {
+    repeated ContinuousNetworkProfilingTaskSampling samplings = 1;
+}
+
+message ContinuousNetworkProfilingTaskSampling {
+    string uriRegex = 1;
+}
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:108]: Merge these two

File: ebpf/profiling/Continuous.proto [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    // current agent contains service and policies
+    repeated ContinuousProfilingServicePolicyQuery policies = 1;
+}
+
+message ContinuousProfilingServicePolicyQuery {
+    // service name of the each process
+    string serviceName = 1;
+    // policy uuid
+    string uuid = 2;
+}
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:49]: Do you have a policy UUID or now? I am confused by following the reply https://github.com/apache/skywalking-data-collect-protocol/pull/83#discussion_r1102531103

Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:49]: Do you have a policy UUID or now? I am confused by following the reply https://github.com/apache/skywalking-data-collect-protocol/pull/83#discussion_r1102531103

Comment (mrproliu) [reply ebpf/profiling/Continuous.proto:49]: One service level policies have one UUID, it used to check whether the service policies that should be updated or not. The mechanism is similar to the `configuration-discovery` in the OAP.

File: ebpf/profiling/Continuous.proto [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+import "common/Command.proto";
+
+service ContinuousProfilingService {
+    // Query continuous profiling policy
+    rpc queryPolicies(ContinuousProfilingPolicyQuery) returns (Commands) {
+    }
+
+    // Report the profiling task when the policy threshold is reached
+    // Use the returned task ID to perform the profiling task, the profiling task reporting same as existing eBPF profiling task report protocol
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:35]:
```suggestion
    // Use the returned task ID to perform the profiling task through EBPFProfilingService#collectProfilingData.
```

File: ebpf/profiling/Continuous.proto [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+message ContinuousProfilingPolicyQuery {
+    // current agent contains service and policies
+    repeated ContinuousProfilingServicePolicyQuery policies = 1;
+}
+
+message ContinuousProfilingServicePolicyQuery {
+    // service name of the each process
+    string serviceName = 1;
+    // policy uuid
+    string uuid = 2;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:49]:
```suggestion
    // UUID represents the version(hash/shasum) of the current policies for the service.
    // This is blank in the initialization stage and is set through the `ContinuousProfilingPolicyQuery` command response for the following rounds of queries.
    string uuid = 2;
```

File: ebpf/profiling/Continuous.proto [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        ContinuousProfilingURICause uri = 3;
+    }
+}
+
+enum ContinuousProfilingCauseType {
+    ProcessCPU = 0;
+    ProcessThreadCount = 1;
+    SystemLoad = 2;
+    ProcessResponseErrorRate = 3;
+    ProcessResponseTimeoutRate = 4;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:85]: Please list the metric meaning/formula.

File: ebpf/profiling/Continuous.proto [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    // over threshold process entity
+    string layer = 1;
+    string serviceName = 2;
+    string instanceName = 3;
+    string processName = 4;
+
+    // reached thresholds causes
+    repeated ContinuousProfilingCause causes = 5;
+
+    // target profiling task execute duration
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:63]:
```suggestion
    // The execution duration for this triggered profiling.
    // This is set at the agent side.
```

File: ebpf/profiling/Continuous.proto [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    oneof cause {
+        ContinuousProfilingSingleValueCause singleValue = 2;
+        ContinuousProfilingURICause uri = 3;
+    }
+}
+
+enum ContinuousProfilingCauseType {
+    // Current process CPU usage percent(0-100).
+    ProcessCPU = 0;
+    // Current process thread count.
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:85]:
```suggestion
    // The number of threads in the process
```

File: ebpf/profiling/Continuous.proto [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+    // Current process CPU usage percent(0-100).
+    ProcessCPU = 0;
+    // The number of threads in the process.
+    ProcessThreadCount = 1;
+    // System load value.
+    SystemLoad = 2;
+    // Process response error rate(0-100).
+    // For the HTTP, the response status [500-600) means error.
+    // Formula: ErrorCount / TotalCount * 100
+    ProcessResponseErrorRate = 3;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:92]:
```suggestion
    // Process response error rate(0-100).
    // HTTP response codes in [500-600) are considered errors.
    // Formula: ErrorCount / TotalCount * 100
    HTTPErrorRate = 3;
```

Comment (wu-sheng) [reply ebpf/profiling/Continuous.proto:92]: Please update command accordingly too.

File: ebpf/profiling/Continuous.proto [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    // Process response average time(ms).
+    // Formula: TotalResponseTime(ms) / TotalCount
+    ProcessResponseAvgTime = 4;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:95]: Is this HTTP specific?

File: ebpf/profiling/Continuous.proto [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+message ContinuousOnCPUProfilingTask {
+}
+
+// eBPF off CPU profiling task
+message ContinuousOffCPUProfilingTask {
+}
+
+// eBPF Network profiling task
+message ContinuousNetworkProfilingTask {
+    string samplingURIRegexes = 1;
```
Comment (wu-sheng) [inline ebpf/profiling/Continuous.proto:122]:
```suggestion
    string samplingURIRegex = 1;
```

This should be one expression?

File: ebpf/profiling/Continuous.proto (hunk 16 above)
Comment (mrproliu) [reply ebpf/profiling/Continuous.proto:95]: For now, we only support HTTP protocol. we could add other protocols if needed.

File: common/Command.proto [hunk 18]
```diff
@@ -74,6 +74,28 @@ import "common/Common.proto";
...
+//         Profiling: Multiple profiling configuration. Map
+//             Key: Profiling type. Enum, value = ON_CPU, OFF_CPU, NETWORK
+//             Value: Profiling policies. Map
+//                 Key: Monitoring type. Enum, value = PROCESS_CPU, PROCESS_THREAD_COUNT, SYSTEM_LOAD, PROCESS_RESPONSE_ERROR_RATE, PROCESS_RESPONSE_AVG_TIME
+//                 Value: Policy configuration. Object.
+//                     Threshold: String
+//                     Period(s): Integer
+//                     Count: Integer
+//                     URIList: List<String>
+//                     URIRegexList: List<String>
```
Comment (wu-sheng) [inline common/Command.proto:93]: https://github.com/apache/skywalking-data-collect-protocol/pull/83/files#r1102879051

This reminds me, Regex should be an expression. We don't need a list, as the regex could represent a list already, right?

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**